FreeRTOS-Plus/Demo/AWS/Device_Defender_Windows_Simulator/Device_Defender_Demo/report_builder.c

```diff
@@ -323,6 +379,185 @@ static eReportBuilderStatus prvWriteConnectionsArray( char * pcBuffer,
 }
 /*-----------------------------------------------------------*/
 
+static eReportBuilderStatus prvWriteCustomMetrics( char * pcBuffer,
+                                                   uint32_t ulBufferLength,
+                                                   const CustomMetric_t * pxCustomMetricsArray,
+                                                   uint32_t ulCustomMetricsArrayLength,
+                                                   uint32_t * pulOutCharsWritten )
+{
+    char * pcCurrentWritePos = pcBuffer;
+    uint32_t i, ulRemainingBufferLength = ulBufferLength;
+    int32_t ulCharactersWritten;
+    eReportBuilderStatus eStatus = eReportBuilderSuccess;
+    const CustomMetric_t * pxCustomMetric;
+
+    configASSERT( pcBuffer != NULL );
+    configASSERT( pxCustomMetricsArray != NULL );
+    configASSERT( pulOutCharsWritten != NULL );
+
+    /* Write the custom metrics array elements. */
+    for( i = 0; i < ulCustomMetricsArrayLength; i++ )
+    {
+        pxCustomMetric = &( pxCustomMetricsArray[ i ] );
+
+        /* Write the metric name key. */
+        ulCharactersWritten = snprintf( pcCurrentWritePos,
+                                        ulRemainingBufferLength,
+                                        "\"%s\":[{",
+                                        pxCustomMetric->pcName );
+
+        if( !reportbuilderSNPRINTF_SUCCESS( ulCharactersWritten, ulRemainingBufferLength ) )
+        {
+            eStatus = eReportBuilderBufferTooSmall;
+            break;
+        }
+        else
+        {
+            ulRemainingBufferLength -= ulCharactersWritten;
+            pcCurrentWritePos += ulCharactersWritten;
+        }
+
+        /* Write the metric data. */
+        if( pxCustomMetric->eType == eCustomMetricNumber )
+        {
+            ulCharactersWritten = snprintf( pcCurrentWritePos,
+                                            ulRemainingBufferLength,
+                                            "\"number\":%lld}],",
+                                            ( long long ) pxCustomMetric->xData.llNumber );
+
+            if( !reportbuilderSNPRINTF_SUCCESS( ulCharactersWritten, ulRemainingBufferLength ) )
+            {
+                eStatus = eReportBuilderBufferTooSmall;
+                break;
+            }
+            else
+            {
+                ulRemainingBufferLength -= ulCharactersWritten;
+                pcCurrentWritePos += ulCharactersWritten;
+            }
+        }
+        else
+        {
+            /* Metric is a list type */
+            if( pxCustomMetric->eType == eCustomMetricNumberList )
+            {
+                ulCharactersWritten = snprintf( pcCurrentWritePos,
+                                                ulRemainingBufferLength,
+                                                "\"number_list\":[" );
```

Comment:
Are these key names for custom metric types not part of the Defender library?

Reply:
They are, but the existing code does not use them. I do agree that it would be better if the regular metrics code was also updated to use defender library constants, especially as the demo hardcodes the long keys for all the keys, instead of using short keys or whichever is configured.